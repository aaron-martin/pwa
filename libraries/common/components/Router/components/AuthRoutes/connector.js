/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { connect } from 'react-redux';
import { isUserLoggedIn } from '../../../../selectors/user';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @return {Object} The extended component props.
 */
const mapStateToProps = state => ({
  isLoggedIn: isUserLoggedIn(state),
});

export default Component => connect(mapStateToProps)(Component);
