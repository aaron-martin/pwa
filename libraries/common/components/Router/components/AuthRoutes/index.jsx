/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import connect from './connector';
import Redirect from '../Redirect';

/**
 * The AuthRoutes component
 * @param {Object} props The props of the component
 * @constructor
 */
const AuthRoutes = props => (
  <div>
    {React.Children.map(props.children, child => (
      props.isLoggedIn
        ? child
        : <Redirect to={props.to} path={child.props.path} />
    ))}
  </div>
);

AuthRoutes.propTypes = {
  children: PropTypes.node.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  isLoggedIn: PropTypes.bool.isRequired,
  to: PropTypes.string.isRequired,
};

AuthRoutes.contextTypes = {
  registerRoute: PropTypes.func.isRequired,
};

export default connect(AuthRoutes);
