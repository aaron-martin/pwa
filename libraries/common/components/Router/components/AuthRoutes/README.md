# AuthRoutes
---

With the AuthRoutes component you can protect routes from being 
opened if the user is not logged in. In this case the user will first be redirected 
to the login page and then forwarded to the actual route after the login was successful. 
If the user is already logged in, the final route will directly show up.

> **Dependents:** <br> `<Route />`

## Getting Started

```
import React from 'react';
import { Route, AuthRoutes } from 'Library/components';

const Routes = () =>
  <App>
    <Route
      path={`${INDEX_PATH}`}
      component="Page"
    />
    <Route
    <AuthRoutes to={`${LOGIN_PATH}`}>
      <Route
        path={`${CHECKOUT_PATH}`}
        component="checkout"
      />
    </AuthRoutes>
  </App>
;
```

## Props

### children (required)

_Type_: `React.Component`<br>

Routes that should be protected by the AuthRoutes component.

###### Usage:

```
<AuthRoutes to={`${LOGIN_PATH}`}>
  <Route
    path={`${CHECKOUT_PATH}`}
    component="checkout"
  />
</AuthRoutes>
```

### isLoggedIn (required)

_Type_: `boolean`<br>

Info about the login state of the current user. Passed via the user connector.

###### Usage:

```
import createHistory from 'history/createBrowserHistory';

const history = createHistory();

<Router history={history}>
  <div>
      <Route path="/category/:categoryId" component={Foo} />
      <Route path="/item/:productId" component={Bar} />
  </div>
</Router>
```

### to (required)

_Type_: `String`<br>

The url to which users should be redirected if they are not logged in.

###### Usage:

```
<AuthRoutes to={`${LOGIN_PATH}`}>
  <Route
    path={`${CHECKOUT_PATH}`}
    component="checkout"
  />
</AuthRoutes>
```
---