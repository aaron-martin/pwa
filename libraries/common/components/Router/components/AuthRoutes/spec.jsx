/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import AuthRoutes from './index';
import MockRoute from '../Route/mock';
import {
  CHECKOUT_PATH,
  LOGIN_PATH,
} from '../../../../constants/RoutePaths';

// Mock the redux connect() method instead of providing a fake store.
jest.mock('./connector', () => obj => obj);

jest.mock('../Redirect/connector', () => (obj) => {
  const newObj = obj;

  newObj.defaultProps = {
    trampolineRedirect: () => {},
  };

  return newObj;
});

describe('<AuthRoutes />', () => {
  let authWrapper;

  beforeEach(() => {
    authWrapper = mount(
      (
        <AuthRoutes to={LOGIN_PATH} isLoggedIn={false}>
          <MockRoute path={CHECKOUT_PATH} component="checkout" />
        </AuthRoutes>
      ), { context: { registerRoute: () => {} } }
    );
  });

  describe('Given the user is logged in', () => {
    beforeEach(() => {
      authWrapper.setProps({ isLoggedIn: true });
    });

    it('should render the Route without any redirect', () => {
      expect(authWrapper.find('MockRoute').exists()).toBe(true);
      expect(authWrapper.find('Redirect').exists()).toBe(false);
      expect(authWrapper).toMatchSnapshot();
    });
  });

  describe('Given the user is logged out', () => {
    beforeEach(() => {
      authWrapper.setProps({ isLoggedIn: false });
    });

    it('should render the redirect', () => {
      expect(authWrapper.find('MockRoute').exists()).toBe(false);
      expect(authWrapper.find('Redirect').exists()).toBe(true);
      expect(authWrapper).toMatchSnapshot();
    });

    it('should pass the correct props to redirect component', () => {
      const props = authWrapper.find('Redirect').props();

      expect(props).toMatchObject({
        to: LOGIN_PATH,
        path: CHECKOUT_PATH,
      });
    });
  });
});
