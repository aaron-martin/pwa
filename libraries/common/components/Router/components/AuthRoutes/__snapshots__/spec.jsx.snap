// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthRoutes /> Given the user is logged in should render the Route without any redirect 1`] = `
<AuthRoutes
  isLoggedIn={true}
  to="/login"
  trampolineRedirect={[Function]}
>
  <div>
    <MockRoute
      component="checkout"
      mockAdd={[Function]}
      mockHide={[Function]}
      mockRemove={[Function]}
      mockShow={[Function]}
      path="/checkout"
    />
  </div>
</AuthRoutes>
`;

exports[`<AuthRoutes /> Given the user is logged out should render the redirect 1`] = `
<AuthRoutes
  isLoggedIn={false}
  to="/login"
  trampolineRedirect={[Function]}
>
  <div>
    <Redirect
      path="/checkout"
      to="/login"
    />
  </div>
</AuthRoutes>
`;
